File: core/internal/storeforward/storeforward.go
// Package storeforward implements Store-and-Forward: queue messages by destination DID,
// flush when the peer is reported online (e.g. after auth).
package storeforward

import (
	"sync"

	"github.com/libp2p/go-libp2p/core/peer"
)

// SendFunc sends a single message payload to the given peer. Called when flushing the queue.
type SendFunc func(pid peer.ID, payload []byte) error

// StoreForward holds pending messages per DID and flushes when a peer is online.
type StoreForward struct {
	mu      sync.Mutex
	pending map[string][][]byte
}

// New returns a new StoreForward.
func New() *StoreForward {
	return &StoreForward{
		pending: make(map[string][][]byte),
	}
}

// Queue adds a message for the given DID. Call when the peer is offline or unknown.
func (s *StoreForward) Queue(did string, payload []byte) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.pending[did] == nil {
		s.pending[did] = [][]byte{}
	}
	s.pending[did] = append(s.pending[did], payload)
}

// FlushForDID sends all queued messages for the given DID to the given peer using sendFn.
// Clears the queue for this DID. Call when the peer is known to be online (e.g. after auth).
func (s *StoreForward) FlushForDID(did string, pid peer.ID, sendFn SendFunc) (sent int, err error) {
	s.mu.Lock()
	msgs := s.pending[did]
	s.pending[did] = nil
	s.mu.Unlock()
	for _, payload := range msgs {
		if sendErr := sendFn(pid, payload); sendErr != nil {
			// Re-queue remaining on first send failure
			s.mu.Lock()
			s.pending[did] = append(msgs[sent:], s.pending[did]...)
			s.mu.Unlock()
			return sent, sendErr
		}
		sent++
	}
	return sent, nil
}

// PendingCount returns the number of queued messages for the DID (or 0).
func (s *StoreForward) PendingCount(did string) int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return len(s.pending[did])
}

File: ts/linkself/src/storeforward.ts
/**
 * Store-and-Forward: queue messages by destination DID, flush when the peer
 * is reported online (e.g. after auth).
 * Port of core/internal/storeforward (Go).
 */

export type SendFunc = (payload: Uint8Array) => Promise<void>;

export class StoreForward {
  private readonly pending = new Map<string, Uint8Array[]>();

  /** Add a message for the given DID. Call when the peer is offline or unknown. */
  queue(did: string, payload: Uint8Array): void {
    const q = this.pending.get(did);
    if (q != null) {
      q.push(payload);
    } else {
      this.pending.set(did, [payload]);
    }
  }

  /**
   * Send all queued messages for the given DID using sendFn and clear the
   * queue. On the first send failure the remaining messages (including any
   * queued concurrently) are re-queued, and the error is rethrown.
   * Returns the number of messages sent.
   */
  async flushForDID(did: string, sendFn: SendFunc): Promise<number> {
    const msgs = this.pending.get(did) ?? [];
    this.pending.delete(did);
    let sent = 0;
    for (const payload of msgs) {
      try {
        await sendFn(payload);
      } catch (err) {
        const remaining = msgs.slice(sent).concat(this.pending.get(did) ?? []);
        this.pending.set(did, remaining);
        throw err;
      }
      sent++;
    }
    return sent;
  }

  /** Number of queued messages for the DID (or 0). */
  pendingCount(did: string): number {
    return this.pending.get(did)?.length ?? 0;
  }
}
